Function-Invokers.adoc

```diff
@@ -0,0 +1,79 @@
+= Riff Function Invokers
+
+== Creating a Function Invoker
+
+=== HTTP
+
+A function invoker must listen on port 8080 for request-response style interactions via a HTTP POST. The HTTP headers have a content-type that can be used to decode the request body.
+
+=== GRPC
+
+A function invoker can also be implemented as a bi-directional stream (as opposed to request-response) in gRPC. The carrier type is a generic `function.Message` defined in the https://github.com/projectriff/function-proto[Function Proto] project. There are language-specific bindings to the `Message` type defined alongside the protobuffer declarations.
+
+== Testing a Function Invoker
+
+For HTTP you should be able to just POST some data to `localhost:8080` and get a response. It is advisable to include a `Content-Type` header, but the function invoker is free to ignore it. The function invoker can make all the decisions it likes about content negotiation, and marshalling the request and response bodies. Any kind of sensible response with 200 OK should be fine - the sidecar takes care of headers in a live Riff system, and copies the response body bytes verbatim.
+
+For gRPC the invoker can still make any decision it likes about content negotiation, and even how many messages to send in response. The function invoker can be tested by writing a tiny gRPC client and sending it a stream of `function.Message`. You would call it a success if you got any kind of messages back, depending on the semantics of the function. If you are supporting request-response, precisely *one* of the response messages must contain the same `correlationId` header as the incoming messages.
+
+== Running the Gateway Locally
+
+Instead of writing custom gRPC test clients, you can run the gateway locally. You will need Kafka, the Riff gateway, a Riff sidecar, and your Function Invoker. Then you can POST messages to the gateway and they will flow through to the function invoker and back through Kafka. This can be useful for end-to-end testing, especially with gRPC function invokers, since there is nothing as convenient as `curl` for ad hoc testing.
+
+One way to run kafka is to use Spring Cloud CLI, or you could use `docker-compose`. Using Spring Cloud would look like this
+
+```
+# once only
+$ sdkman install springboot 1.5.10.RELEASE
```

Comment:
I don't have sdkman - maybe we should provide docker-compose instructions, most developers would already have that